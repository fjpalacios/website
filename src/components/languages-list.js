import * as React from 'react'
import './languages-list.scss'

const LanguagesList = ({ data }) => {
  return (
    <>
      {data.map((item, key) => {
        const Name = () => {
          if (item.url) {
            return <a href={item.url}>{item.name}</a>
          }

          return <span>{item.name}</span>
        }

        const Dates = () => {
          if (item.dates) {
            return <div className="languages-list__dates">{item.dates}</div>
          }

          return <></>
        }

        return (
          <div key={key} className="languages-list">
            <div className="languages-list__name">
              <Name />
            </div>
            <Dates />
            <ul className="languages-list__languages">
              {item.languages.map((language, key) => {
                return (
                  <li key={key}>
                    <a href={`https://ddg.gg/?q=${language}`}>{`#${language}`}</a>
                  </li>
                )
              })}
            </ul>
            <div className="languages-list__desc" dangerouslySetInnerHTML={{ __html: item.desc }}></div>
          </div>
        )
      })}
    </>
  )
}

export default LanguagesList
