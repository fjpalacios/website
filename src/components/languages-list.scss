.languages-list {
  display: grid;
  grid-template-areas:
    'name dates'
    'languages languages'
    'desc desc';
  margin-bottom: 20px;
  padding: 0 20px;

  @include small-and-up {
    grid-template-areas:
      'name dates'
      'languages .'
      'desc .';
  }

  @include print {
    margin-bottom: 10px;
    padding: 0 10px;
  }

  &__name {
    grid-area: name;
    text-transform: uppercase;
  }

  &__dates {
    display: grid;
    grid-area: dates;
    justify-content: end;
  }

  &__languages {
    display: flex;
    grid-area: languages;
    margin: 0;
    padding: 0;

    li {
      display: inline-block;
      padding-right: 10px;
    }
  }

  &__desc {
    grid-area: desc;
  }
}
