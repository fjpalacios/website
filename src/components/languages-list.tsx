import './languages-list.scss'
import React, { FunctionComponent, ReactElement } from 'react'

export type languagesList = {
  name: string
  url?: string
  dates?: string
  desc: string
  languages: string[]
}

type languagesListProps = {
  data: languagesList[]
}

export const LanguagesList: FunctionComponent<languagesListProps> = ({
  data,
}): ReactElement => {
  return (
    <>
      {data.map((item, key) => {
        const Name = (): ReactElement => {
          if (item.url) {
            return <a href={item.url}>{item.name}</a>
          }

          return <span>{item.name}</span>
        }

        const Dates = (): ReactElement => {
          if (item.dates) {
            return <div className="languages-list__dates">{item.dates}</div>
          }

          return <></>
        }

        return (
          <div key={key} className="languages-list">
            <div className="languages-list__name">
              <Name />
            </div>
            <Dates />
            <ul className="languages-list__languages">
              {item.languages.map((language, key) => {
                return (
                  <li key={key}>
                    <a
                      href={`https://ddg.gg/?q=${language}`}
                    >{`#${language}`}</a>
                  </li>
                )
              })}
            </ul>
            <div
              className="languages-list__desc"
              dangerouslySetInnerHTML={{ __html: item.desc }}
            ></div>
          </div>
        )
      })}
    </>
  )
}
