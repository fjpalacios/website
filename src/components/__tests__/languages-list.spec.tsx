import { languagesList, LanguagesList } from '../languages-list'
import createComponentWithIntl from '../../../__helpers__/i18n-component'
import Faker from 'faker'
import React from 'react'

describe('LanguagesList', () => {
  Faker.seed(123)
  const data: languagesList[] = [
    {
      name: Faker.internet.domainName(),
      desc: Faker.lorem.sentence(),
      languages: [Faker.hacker.verb()],
    },
  ]

  it('should render properly', () => {
    const languagesList = createComponentWithIntl(<LanguagesList data={data} />)
    expect(languagesList).toMatchSnapshot()
  })

  it('should show the name', () => {
    const languagesList = createComponentWithIntl(<LanguagesList data={data} />)
    const name = languagesList.find('.languages-list__name span').text()
    expect(name).toBe(data[0].name)
  })

  it('should show a url if it exists', () => {
    data[0].url = 'https://duckduckgo.com'
    const languagesList = createComponentWithIntl(<LanguagesList data={data} />)
    const name = languagesList.find('.languages-list__name a').text()
    expect(name).toBe(data[0].name)
    data[0].url = undefined
  })

  it('should show a list of languages', () => {
    const languagesList = createComponentWithIntl(<LanguagesList data={data} />)
    const location = languagesList
      .find('.languages-list__languages li a')
      .text()
    expect(location).toBe(`#${data[0].languages[0]}`)
  })

  it('should not show the dates if they are not defined', () => {
    const languagesList = createComponentWithIntl(<LanguagesList data={data} />)
    const dates = languagesList.find('.languages-list__dates')
    expect(dates).toBeEmpty
  })

  it('should show the dates if they are defined', () => {
    data[0].dates = Faker.date.past().toDateString()
    const languagesList = createComponentWithIntl(<LanguagesList data={data} />)
    const dates = languagesList.find('.languages-list__dates').text()
    expect(dates).toBe(data[0].dates)
    data[0].dates = undefined
  })

  it('should show the description', () => {
    const languagesList = createComponentWithIntl(<LanguagesList data={data} />)
    const desc = languagesList.find('.languages-list__desc').text()
    expect(desc).toBe(data[0].desc)
  })
})
