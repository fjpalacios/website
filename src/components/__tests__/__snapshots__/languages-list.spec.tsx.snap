// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`LanguagesList should render properly 1`] = `
<IntlProvider
  defaultFormats={Object {}}
  defaultLocale="en"
  formats={Object {}}
  locale="en"
  messages={
    Object {
      "about": Object {
        "education": "Education",
        "experience": "Experience",
        "freelance": "Freelance work",
        "me": "About me",
        "projects": "Side projects",
        "talks": "Talks / workshops",
        "volunteering": "Volunteering",
      },
      "subtitle": "Software Developer",
      "title": "Francisco Javier Palacios PÃ©rez",
    }
  }
  onError={[Function]}
  textComponent={Symbol(react.fragment)}
>
  <LanguagesList
    data={
      Array [
        Object {
          "desc": "Molestiae tempora eligendi omnis quisquam.",
          "languages": Array [
            "synthesize",
          ],
          "name": "maurine.net",
        },
      ]
    }
  >
    <div
      className="languages-list"
      key="0"
    >
      <div
        className="languages-list__name"
      >
        <Name>
          <span>
            maurine.net
          </span>
        </Name>
      </div>
      <Dates />
      <ul
        className="languages-list__languages"
      >
        <li
          key="0"
        >
          <a
            href="https://ddg.gg/?q=synthesize"
          >
            #synthesize
          </a>
        </li>
      </ul>
      <div
        className="languages-list__desc"
        dangerouslySetInnerHTML={
          Object {
            "__html": "Molestiae tempora eligendi omnis quisquam.",
          }
        }
      />
    </div>
  </LanguagesList>
</IntlProvider>
`;
